metricbeat/mb/builders.go

```diff
@@ -0,0 +1,163 @@
+package mb
+
+import (
+	"fmt"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/joeshaw/multierror"
+)
+
+var debugf = logp.MakeDebug("mb")
+
+// NewModules builds new Modules and their associated MetricSets based on the
+// provided configuration data. config is a list module config data (the data
+// will be unpacked into ModuleConfig structs). r is the Register where the
+// ModuleFactory's and MetricSetFactory's will be obtained from. This method
+// returns a mapping of Modules to MetricSets or an error.
+func NewModules(config []*common.Config, r *Register) (map[Module][]MetricSet, error) {
+	baseModules, err := newBaseModulesFromConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create new Modules using the registered ModuleFactory's
+	modules := make([]Module, 0, len(baseModules))
+	var errs multierror.Errors
+	for _, bm := range baseModules {
+		f := r.moduleFactory(bm.Name())
+		if f == nil {
+			f = DefaultModuleFactory
+		}
+
+		module, err := f(bm)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		modules = append(modules, module)
+	}
+
+	if errs.Err() != nil {
+		return nil, errs.Err()
+	}
+
+	// Create new MetricSets for each Module using the registered MetricSetFactory's
+	modToMetricSets := make(map[Module][]MetricSet, len(modules))
+	for _, bms := range newBaseMetricSets(modules) {
+		f, err := r.metricSetFactory(bms.Module().Name(), bms.Name())
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		metricSet, err := f(bms)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		err = mustImplementFetcher(metricSet)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		err = mustHaveModule(metricSet, bms)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if list, ok := modToMetricSets[metricSet.Module()]; ok {
+			list = append(list, metricSet)
+			modToMetricSets[metricSet.Module()] = list
+			continue
+		}
+		modToMetricSets[metricSet.Module()] = []MetricSet{metricSet}
+	}
+
+	if errs.Err() != nil {
+		return nil, errs.Err()
```

Comment:
Same question here if we should use handle the conversion upstream instead of here?